tools/codegen/api/types.py

```diff
@@ -40,15 +53,217 @@ def __str__(self) -> str:
+    # Return this argument pack, but with default stripped
+    def no_default(self: _T) -> _T:
+        raise NotImplementedError
+
+    # Unpack the pack into a sequence of arguments, discarding
+    # semantic information, and also discarding the implicit this
+    # argument that doesn't actually show up in declarations
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        raise NotImplementedError
+
+# Lifts a single CppArgument into a pack.
+@dataclass(frozen=True)
+class CppSingleArgumentPack(CppArgumentPackIface):
+    this: CppArgument
+
+    def no_default(self) -> 'CppSingleArgumentPack':
+        return CppSingleArgumentPack(self.this.no_default())
+
+    @property
+    def type(self) -> str:
+        return self.this.type
+
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return [self.this]
+
+# Describe an implicit this argument (*this) on methods in the C++ API.
+# We don't use CppSingleArgumentPack because these never show up
+# in the explicit arguments list
+@dataclass(frozen=True)
+class CppThisArgumentPack(CppArgumentPackIface):
+    # The grouped JIT argument this formal was derived from
+    argument: ThisArgument
+
+    # C++ type, e.g., Tensor&
+    type: str
+
+    # this arguments are never defaulted
+    def no_default(self) -> 'CppThisArgumentPack':
+        return self
+
+    # The this argument is implicit, so it's not included in the
+    # explicit arguments list.
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return []
+
+# Semantically represents a bundle of CppArguments that collectively
+# represent a TensorOptions.  If you don't care about TensorOptions
+# processing, think of this as just a list of four CppArguments; however
+# if you need to bundle these arguments back into a single
+# TensorOptions, it will be easiest to operate on this struct as a
+# whole.
+#
+# NOTE: this does NOT represent a 'const TensorOptions&' argument.
+# If you have one of those, it will be CppSingleArgumentPack
+@dataclass(frozen=True)
+class CppTensorOptionsArgumentPack(CppArgumentPackIface):
+    argument: TensorOptionsArguments
+    dtype: CppArgument
+    layout: CppArgument
+    device: CppArgument
+    pin_memory: CppArgument
+
+    # Remove the defaults from each of the constituent arguments
+    # representing the TensorOptions
+    def no_default(self) -> 'CppTensorOptionsArgumentPack':
+        return CppTensorOptionsArgumentPack(
+            argument=self.argument,
+            dtype=self.dtype.no_default(),
+            layout=self.layout.no_default(),
+            device=self.device.no_default(),
+            pin_memory=self.pin_memory.no_default(),
+        )
+
+    # Flatten the TensorOptions into individual CppArguments
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return [self.dtype, self.layout, self.device, self.pin_memory]
+
+# Use this instead of CppArgumentPackIface, as this is a closed union
+CppArgumentPack = Union[
```

Comment:
Just to recap the common theme from a few of the preceding comments - given that representing a signature as (so to speak) `(arg | this_arg | tensor_opts_arg)*` is really very loose, I'd be tempted to do away with this union and see what the architecture looks like after fixing all the places where it currently appears. Mostly I'd anticipate certain decisions that are currently based on type sniffing would be hoisted into `CppSignature` and be based on the structure of the signature.

Reply:
Ah, there is a reason we can't do this. To restate your proposal, your thinking is that we should have:

```
args: Sequence[CppArgument]
this_arg: Optional[CppThisArgument]
tensor_opts_arg: Optional[CppTensorOptsArgument]
```

(and also same structure in the non-cpp model. This prevents us from having to do a pattern match. The reason this doesn't work is that this representation underspecifies the relative locations of all the arguments with each other. There are two major underspecifications:

* The this/self argument isn't always at the beginning of the function. where is the most well-known counterexample:

```
- func: where.self(Tensor condition, Tensor self, Tensor other) -> Tensor
  use_c10_dispatcher: full
  variants: function, method
```

* The TensorOptions arguments isn't always at the end of the function. In fact, ALL factory functions that accept memory format violate this condition (because memory format is not considered part of the grouping at the moment):

```
- func: empty.memory_format(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
```

Now, they do say exceptions shoudn't write the law, and so maybe we should manually work around these exceptions to make your proposal work out. But this was at least some of the motivation to set things up this way, at least initially.

Reply:
Yeah, I realize that's not quite enough - I was thinking something more like
```
args: Sequence[CppArgument]
this_arg: Optional[Tuple[CppThisArgument, int]]
tensor_opts_arg: Optional[Tuple[CppTensorOptsArgument, int]]
```

Reply:
Index splicing is tricky because you need to do the splicing "all in one go" otherwise things get shifted as you splice each one in. It might be better to just have "before" and "after" argument lists.